File: WayforkWindows/app/lib/core/app/log_file.dart
import 'dart:convert';
import 'dart:io';

import 'package:wayfork/core/ipc/payloads.dart';
import 'package:wayfork/core/model/settings.dart';

/// One line of `runtime.log` / `wayfork.log` (docs/design/06-logging.md).
abstract final class LogLineFormat {
  /// ISO-8601 UTC with milliseconds, source, upper-cased level, message.
  static String format(LogLine line) {
    final message = line.message.replaceAll(RegExp(r'[\r\n]'), ' ');
    return '${timestamp(line.ts)} ${line.source} '
        '${line.level.jsonValue.toUpperCase()} $message';
  }

  /// `2026-08-25T12:00:00.123Z`.
  static String timestamp(DateTime date) {
    final utc = date.toUtc();
    String pad(int value, int width) => value.toString().padLeft(width, '0');
    return '${pad(utc.year, 4)}-${pad(utc.month, 2)}-${pad(utc.day, 2)}T'
        '${pad(utc.hour, 2)}:${pad(utc.minute, 2)}:${pad(utc.second, 2)}.'
        '${pad(utc.millisecond, 3)}Z';
  }

  /// Parses a line produced by [format]; malformed and non-canonical lines
  /// return null.
  static LogLine? parse(String text) {
    if (text.isEmpty || text.contains('\n') || text.contains('\r')) return null;
    final parts = _split(text, 3);
    if (parts.length != 4 ||
        parts[0].isEmpty ||
        parts[1].isEmpty ||
        parts[2].isEmpty) {
      return null;
    }
    final DateTime date;
    try {
      date = DateTime.parse(parts[0]).toUtc();
    } on FormatException {
      return null;
    }
    if (timestamp(date) != parts[0]) return null;
    final levelText = parts[2];
    if (levelText != levelText.toUpperCase()) return null;
    final LogLevel level;
    try {
      level = LogLevel.fromJson(levelText.toLowerCase());
    } on FormatException {
      return null;
    }
    return LogLine(ts: date, source: parts[1], level: level, message: parts[3]);
  }

  /// Splits on single spaces at most [maxSplits] times, keeping empty pieces.
  static List<String> _split(String text, int maxSplits) {
    final parts = <String>[];
    var start = 0;
    while (parts.length < maxSplits) {
      final index = text.indexOf(' ', start);
      if (index < 0) break;
      parts.add(text.substring(start, index));
      start = index + 1;
    }
    parts.add(text.substring(start));
    return parts;
  }
}

enum AppLogFileError { invalidConfiguration, openFailed }

final class AppLogFileException implements Exception {
  const AppLogFileException(this.kind, {this.message});

  final AppLogFileError kind;
  final String? message;

  @override
  String toString() =>
      message == null ? 'AppLogFileException($kind)' : '$kind: $message';
}

/// Append-only app log with timestamped size rotation and age-based retention.
/// The directory lives under `%LOCALAPPDATA%`, which is private to the user by
/// its inherited ACL, so no explicit permissions are set.
final class AppLogFile {
  AppLogFile(this.directory, this.name, {this.maxBytes = defaultMaxBytes})
    : file = File(_join(directory.path, '$name.log')) {
    if (name.isEmpty || maxBytes <= 0) {
      throw const AppLogFileException(AppLogFileError.invalidConfiguration);
    }
    try {
      directory.createSync(recursive: true);
      _open();
    } on FileSystemException catch (error) {
      throw AppLogFileException(
        AppLogFileError.openFailed,
        message: error.message,
      );
    }
  }

  static const defaultMaxBytes = 5 * 1024 * 1024;

  final Directory directory;
  final String name;
  final int maxBytes;
  final File file;

  RandomAccessFile? _handle;
  int _byteCount = 0;
  bool _isClosed = false;

  void append(String text) {
    if (_isClosed) return;
    final bytes = utf8.encode('$text\n');
    if (_byteCount > 0 && _byteCount + bytes.length > maxBytes) _rotate();
    final handle = _handle;
    if (handle == null) return;
    try {
      handle.writeFromSync(bytes);
      _byteCount += bytes.length;
    } on FileSystemException {
      // A log line that cannot be written is dropped, like the macOS app.
    }
  }

  /// The last [maxLines] lines, oldest first, reading at most [maxBytes] from
  /// the end of the current file.
  List<String> tail(int maxLines, {int maxBytes = defaultMaxBytes}) {
    if (maxLines <= 0 || maxBytes <= 0 || _isClosed) return const [];
    final RandomAccessFile reader;
    try {
      reader = file.openSync();
    } on FileSystemException {
      return const [];
    }
    try {
      final size = reader.lengthSync();
      final length = size < maxBytes ? size : maxBytes;
      if (length <= 0) return const [];
      final start = size - length;
      reader.setPositionSync(start);
      var bytes = reader.readSync(length);
      if (start > 0) {
        reader.setPositionSync(start - 1);
        final previous = reader.readSync(1);
        if (previous.isEmpty || previous[0] != 0x0A) {
          final newline = bytes.indexOf(0x0A);
          if (newline < 0) return const [];
          bytes = bytes.sublist(newline + 1);
        }
      }
      final lines = utf8.decode(bytes, allowMalformed: true).split('\n');
      if (lines.isNotEmpty && lines.last.isEmpty) lines.removeLast();
      return lines.length > maxLines
          ? lines.sublist(lines.length - maxLines)
          : lines;
    } on FileSystemException {
      return const [];
    } finally {
      reader.closeSync();
    }
  }

  void close() {
    if (_isClosed) return;
    _isClosed = true;
    _closeHandle();
  }

  /// Rotated files of [name] in [directory], oldest first by name.
  static List<File> rotatedFiles(Directory directory, String name) {
    final prefix = '$name-';
    if (!directory.existsSync()) return const [];
    final files = directory.listSync().whereType<File>().where((file) {
      final base = _basename(file.path);
      return base.startsWith(prefix) && base.endsWith('.log');
    }).toList();
    files.sort((a, b) => _basename(a.path).compareTo(_basename(b.path)));
    return files;
  }

  /// Deletes rotated files older than [retentionDays] (all of them when the
  /// retention is zero or negative); returns what was deleted.
  static List<File> prune(
    Directory directory,
    String name, {
    required int retentionDays,
    DateTime? now,
  }) {
    final cutoff = (now ?? DateTime.now()).subtract(
      Duration(days: retentionDays),
    );
    final deleted = <File>[];
    for (final file in rotatedFiles(directory, name)) {
      bool shouldDelete;
      if (retentionDays <= 0) {
        shouldDelete = true;
      } else {
        try {
          shouldDelete = file.lastModifiedSync().isBefore(cutoff);
        } on FileSystemException {
          shouldDelete = false;
        }
      }
      if (!shouldDelete) continue;
      try {
        file.deleteSync();
        deleted.add(file);
      } on FileSystemException {
        // Keep going; a locked file is retried on the next prune.
      }
    }
    return deleted;
  }

  void _open() {
    final handle = file.openSync(mode: FileMode.append);
    _handle = handle;
    _byteCount = handle.lengthSync();
  }

  void _closeHandle() {
    final handle = _handle;
    _handle = null;
    if (handle == null) return;
    try {
      handle.closeSync();
    } on FileSystemException {
      // Nothing left to do with a handle that refuses to close.
    }
  }

  void _rotate() {
    _closeHandle();
    try {
      file.renameSync(_availableRotatedPath());
    } on FileSystemException {
      // Rename failed: keep appending to the current file.
    }
    try {
      _open();
    } on FileSystemException {
      _byteCount = 0;
    }
  }

  String _availableRotatedPath() {
    final now = DateTime.now().toUtc();
    String pad(int value, int width) => value.toString().padLeft(width, '0');
    final stem =
        '$name-${pad(now.year, 4)}${pad(now.month, 2)}${pad(now.day, 2)}-'
        '${pad(now.hour, 2)}${pad(now.minute, 2)}${pad(now.second, 2)}';
    var candidate = _join(directory.path, '$stem.log');
    var suffix = 1;
    while (File(candidate).existsSync()) {
      candidate = _join(directory.path, '$stem-$suffix.log');
      suffix += 1;
    }
    return candidate;
  }

  static String _join(String directory, String name) =>
      directory.endsWith(Platform.pathSeparator)
      ? '$directory$name'
      : '$directory${Platform.pathSeparator}$name';

  static String _basename(String path) {
    final index = path.lastIndexOf(RegExp(r'[\\/]'));
    return index < 0 ? path : path.substring(index + 1);
  }
}

File: macos/WayforkCore/Sources/WayforkCore/App/LogFile.swift
import Darwin
import Foundation

/// One line of `runtime.log` / `wayfork.log`.
public enum LogLineFormat {
    /// `ISO8601DateFormatter` is documented as thread-safe; one instance for every line.
    nonisolated(unsafe) private static let formatter: ISO8601DateFormatter = {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        return formatter
    }()

    /// ISO-8601 UTC with milliseconds, source, upper-cased level, message.
    public static func format(_ line: LogLine) -> String {
        let message = line.message.map { $0.isNewline ? " " : String($0) }.joined()
        return "\(formatter.string(from: line.ts)) \(line.source) "
            + "\(line.level.rawValue.uppercased()) \(message)"
    }

    /// Parses a line produced by `format`; malformed and non-canonical lines return `nil`.
    public static func parse(_ text: String) -> LogLine? {
        guard !text.isEmpty, !text.contains(where: \Character.isNewline) else {
            return nil
        }
        let parts = text.split(separator: " ", maxSplits: 3, omittingEmptySubsequences: false)
        guard parts.count == 4, !parts[0].isEmpty, !parts[1].isEmpty, !parts[2].isEmpty else {
            return nil
        }

        let timestamp = String(parts[0])
        guard let date = formatter.date(from: timestamp), formatter.string(from: date) == timestamp
        else {
            return nil
        }
        let levelText = String(parts[2])
        guard
            levelText == levelText.uppercased(),
            let level = LogLevel(rawValue: levelText.lowercased())
        else {
            return nil
        }
        return LogLine(ts: date, source: String(parts[1]), level: level, message: String(parts[3]))
    }
}

public enum AppLogFileError: Error, Sendable, Equatable {
    case invalidConfiguration
    case openFailed(errno: Int32)
}

/// Append-only app log with timestamped size rotation and age-based retention.
public final class AppLogFile: @unchecked Sendable {
    public let url: URL

    private let directory: URL
    private let name: String
    private let maxBytes: Int
    private let lock = NSLock()
    private var fileDescriptor: Int32
    private var byteCount: Int
    private var isClosed = false

    public init(directory: URL, name: String, maxBytes: Int = 5 * 1024 * 1024) throws {
        guard !name.isEmpty, maxBytes > 0 else {
            throw AppLogFileError.invalidConfiguration
        }
        try FileManager.default.createDirectory(
            at: directory,
            withIntermediateDirectories: true,
            attributes: [.posixPermissions: 0o700]
        )
        guard chmod(directory.path, 0o700) == 0 else {
            throw AppLogFileError.openFailed(errno: errno)
        }

        let url = directory.appendingPathComponent("\(name).log")
        let opened = try Self.openLog(at: url.path)
        self.directory = directory
        self.name = name
        self.maxBytes = maxBytes
        self.url = url
        fileDescriptor = opened.fd
        byteCount = opened.size
    }

    deinit {
        close()
    }

    public func append(_ text: String) {
        let bytes = Array((text + "\n").utf8)
        lock.withLock {
            guard !isClosed else { return }
            if byteCount > 0, byteCount + bytes.count > maxBytes {
                rotate()
            }
            guard fileDescriptor >= 0 else { return }

            var offset = 0
            while offset < bytes.count {
                let written = bytes.withUnsafeBytes { buffer in
                    Darwin.write(
                        fileDescriptor,
                        buffer.baseAddress!.advanced(by: offset),
                        bytes.count - offset
                    )
                }
                if written > 0 {
                    offset += written
                    byteCount += written
                } else if written == -1, errno == EINTR {
                    continue
                } else {
                    return
                }
            }
        }
    }

    public func tail(maxLines: Int, maxBytes: Int = 5 * 1024 * 1024) -> [String] {
        guard maxLines > 0, maxBytes > 0 else { return [] }
        return lock.withLock {
            guard !isClosed, fileDescriptor >= 0 else { return [] }

            var info = stat()
            guard fstat(fileDescriptor, &info) == 0 else { return [] }
            let length = min(Int(info.st_size), maxBytes)
            guard length > 0 else { return [] }
            let start = Int(info.st_size) - length
            var bytes = [UInt8](repeating: 0, count: length)
            var offset = 0
            while offset < length {
                let count = bytes.withUnsafeMutableBytes { buffer in
                    pread(
                        fileDescriptor,
                        buffer.baseAddress!.advanced(by: offset),
                        length - offset,
                        off_t(start + offset)
                    )
                }
                if count > 0 {
                    offset += count
                } else if count == -1, errno == EINTR {
                    continue
                } else {
                    break
                }
            }
            bytes.removeSubrange(offset..<bytes.count)
            if start > 0, !Self.startsAtLineBoundary(fileDescriptor: fileDescriptor, start: start) {
                guard let newline = bytes.firstIndex(of: 0x0A) else { return [] }
                bytes.removeSubrange(...newline)
            }

            var lines = String(decoding: bytes, as: UTF8.self)
                .split(separator: "\n", omittingEmptySubsequences: false)
                .map(String.init)
            if lines.last == "" { lines.removeLast() }
            return Array(lines.suffix(maxLines))
        }
    }

    public func close() {
        lock.withLock {
            guard !isClosed else { return }
            isClosed = true
            if fileDescriptor >= 0 {
                Darwin.close(fileDescriptor)
                fileDescriptor = -1
            }
        }
    }

    public static func rotatedFiles(directory: URL, name: String) -> [URL] {
        let prefix = "\(name)-"
        return
            (try? FileManager.default.contentsOfDirectory(
                at: directory,
                includingPropertiesForKeys: nil
            ))?.filter {
                $0.lastPathComponent.hasPrefix(prefix) && $0.pathExtension == "log"
            }.sorted { $0.lastPathComponent < $1.lastPathComponent } ?? []
    }

    @discardableResult
    public static func prune(
        directory: URL,
        name: String,
        retentionDays: Int,
        now: Date = Date()
    ) -> [URL] {
        let cutoff = now.addingTimeInterval(-Double(retentionDays) * 86_400)
        return rotatedFiles(directory: directory, name: name).filter { url in
            let shouldDelete: Bool
            if retentionDays <= 0 {
                shouldDelete = true
            } else {
                let values = try? url.resourceValues(forKeys: [.contentModificationDateKey])
                shouldDelete = values?.contentModificationDate.map { $0 < cutoff } ?? false
            }
            guard shouldDelete, (try? FileManager.default.removeItem(at: url)) != nil else {
                return false
            }
            return true
        }
    }

    private func rotate() {
        if fileDescriptor >= 0 {
            Darwin.close(fileDescriptor)
            fileDescriptor = -1
        }

        let rotatedURL = availableRotatedURL()
        guard rename(url.path, rotatedURL.path) == 0 else {
            reopenCurrentFile()
            return
        }
        reopenCurrentFile()
    }

    private func availableRotatedURL() -> URL {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.calendar = Calendar(identifier: .gregorian)
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        formatter.dateFormat = "yyyyMMdd-HHmmss"
        let stem = "\(name)-\(formatter.string(from: Date()))"
        var candidate = directory.appendingPathComponent("\(stem).log")
        var suffix = 1
        while FileManager.default.fileExists(atPath: candidate.path) {
            candidate = directory.appendingPathComponent("\(stem)-\(suffix).log")
            suffix += 1
        }
        return candidate
    }

    private func reopenCurrentFile() {
        guard let opened = try? Self.openLog(at: url.path) else {
            byteCount = 0
            return
        }
        fileDescriptor = opened.fd
        byteCount = opened.size
    }

    private static func openLog(at path: String) throws -> (fd: Int32, size: Int) {
        let fd = Darwin.open(path, O_RDWR | O_CREAT | O_APPEND | O_CLOEXEC, 0o600)
        guard fd >= 0 else { throw AppLogFileError.openFailed(errno: errno) }
        guard fchmod(fd, 0o600) == 0 else {
            let savedErrno = errno
            Darwin.close(fd)
            throw AppLogFileError.openFailed(errno: savedErrno)
        }
        var info = stat()
        guard fstat(fd, &info) == 0 else {
            let savedErrno = errno
            Darwin.close(fd)
            throw AppLogFileError.openFailed(errno: savedErrno)
        }
        return (fd, Int(info.st_size))
    }

    private static func startsAtLineBoundary(fileDescriptor: Int32, start: Int) -> Bool {
        var previous: UInt8 = 0
        return pread(fileDescriptor, &previous, 1, off_t(start - 1)) == 1 && previous == 0x0A
    }
}
